### ISSUE TITLE
how to change user groups in octobercms automatically after payment

### ISSUE THREAD

**Issue by Mins17** [rainlab/user-plugin#352]
Hi, I am using "user plugin". In my website I have some paid content pages. The pages are accessible to only those users who belong to user-group -"paid". 
I want to change the user group of users to "paid" automatically, once they make the payment.
Presently the only way its working is
1. I manually set a user as 'paid' in the backend.
2. I check the if the user is in paid group by- {% if user.groups[0].id == 3 %}
3. So the user whom I set as paid gets the access, the rest of the users don't.   
4. But i am seeking answer to make it automatic once the payment is done.
How can i achieve this. Thanks in advance.

**Comment by LukeTowers** [comment]
```php
Event::listen('myvendor.myplugin.payment.completed', function ($user) {
    $group = \RainLab\User\Models\UserGroup::where('code', 'paid')->first();
    $user->groups()->add($group);
    $user->save();
});
```

**Comment by Mins17** [comment]
Thanks Luke. I am quite new to october cms and coding. Can you please elaborate the answer. I mean in which files I have to make these changes. I know you've scarcity of time. But this will help me a lot. Thanks.

**Comment by LukeTowers** [comment]
@Mins17 you'll need to create your own plugin and then you can place that code in your `boot()` method of your `Plugin.php` class. See https://octobercms.com/docs/plugin/registration and https://octobercms.com/docs/plugin/registration#registration-methods. You'll need to import the `Event` class at the top of the file with `use Event;`, and if you get stuck on any other of the PHP coding parts try googling first and then ask around the slack, usually people are pretty willing to help out there.

**Comment by Mins17** [comment]
Thanks Luke. Appreciate your help.

**Comment by Corymillz** [comment]
@lukeTowers. Please I need some help here on how to solve this. I followed your code above on adding users to a groups automatically. I have two registration forms for two user groups and used

```Event::listen('rainlab.user.activate, function...``` and pretty much every other thing there

My question is how do i indicate in these forms which group to send its contents to. I've tried changing the paramCode to "buyer-account" which is one of the names of my group but it isn't adding to the group. I have been on this for hours. How do i get this working?

**Comment by bennothommo** [comment]
@Corymillz It might be better to ask your questions in one of the avenues below. GitHub issues isn't the best place for questions like this.

Forum: https://octobercms.com/forum
StackOverflow: https://stackoverflow.com/questions/tagged/octobercms
IRC: #october on Freenode or https://octobercms.com/chat
Slack: https://octobercms-slack.herokuapp.com/

**Comment by Corymillz** [comment]
@bennothommo i know and i have tried those places without results before coming here.

**Comment by bennothommo** [comment]
@Corymillz If you have two registration forms which assign users to two separate groups, your best bet is to use the `rainlab.user.register` event, as its first parameter (`$user`) is the User model and the second paramater (`$data`) is the POST data from the registration form - for example, you could have a hidden field in the registration form which is different depending on which form you are using, and thus could change the user's group depending on this value.

**Comment by LukeTowers** [comment]
You could also look at interacting with the global request object directly to identify what URL the form was submitted on (something like `Request::url()` I believe)

**Comment by Corymillz** [comment]
OK. I am currently trying to make at least one form add users to a group first. This is exactly what I have done. I created a new plugin. In the Plugin.php file I have ```use Event``` at the top of the file

I created a boot method and have this in it.

```
   public function boot()
{ 
  $group = 
    \Rainlab\User\Models\UserGroup::where('code', 'seller')->first();
    

    $user->groups()->add($group);
    $user->save(); 
}); 
}
```
I think that should be enough if I am right to add users to an already created group but it does not work? What am I missing or doing wrong?

**Comment by LukeTowers** [comment]
@Corymillz please post your full plugin.php file here for review

**Comment by Corymillz** [comment]
@LukeTowers  This is all I have in the Plugin.php
```
<?php namespace Corymillz\Newplug;

use System\Classes\PluginBase;
use Event;

class Plugin extends PluginBase
{
   public function registerComponents()
    {
    }

    public function registerSettings()
    {
    }

   public function boot()
{ 
Event::listen('rainlab.user.register', function ($user) {
  $group = 
    \Rainlab\User\Models\UserGroup::where('code', 'seller')->first();
    

    $user->groups()->add($group);
    $user->save(); 
}); 
}

}
```

**Comment by bennothommo** [comment]
@Corymillz I've tested your code and it does work. It sounds like you haven't configured the plugin right if it is not running the event and removing it from the plugins list.

You need to make sure you have both a `Plugin.php` file in the plugin root folder, and an `updates/version.yaml` file with at least one version starting from 1.0 (ie. `1.0.0: Initialised plugin.`). You should then be able to click the *Check for updates* button in the *Updates & Plugins* screen in the backend and see your plugin in the list.

**Comment by LukeTowers** [comment]
@Corymillz also make sure that your plugin.php file is located in `plugins/corymillz/newplug/Plugin.php`

**Comment by Corymillz** [comment]
Yeah. No longer removing from the list. In the backend, I've already created the group and its code in the UserGroup. All fields in the registration shows up as it should. With all these in place the user is suppose to add up in the group on registration. What other checks do I need to make?

**Comment by bennothommo** [comment]
@Corymillz `Request::url()` would be used as part of your event listener. With it, you could check the URL that the registration came from as part of the `rainlab.user.activate` event.

**Comment by Corymillz** [comment]
Doesn't seem to work  or I really don't know what I am doing. Places I've tried it broke the code.

**Comment by LukeTowers** [comment]
that wouldn't work at all. `Request::url()` returns a string, so you could try doing the following:

```php
$groupCode = false;
if (ends_with(Request::url(), '/seller-registration')) {
    $groupCode = 'seller';
} elseif (ends_with(Request::url(), '/buyer-registration')) {
    $groupCode = 'buyer';
}

if ($group = UserGroup::where('code', $groupCode)->first()) {
    Event::listen('rainlab.user.register', function ($user, $data) use ($group) {
        $user->groups()->add($group);
        $user->save();
    });
}
```

**Comment by Corymillz** [comment]
Thanks guys for putting me in the right direction. Finally got it to work using this just in case somebody else needs it.

```
  public function boot()
{ 

if (\Request::is('seller-registration')){

Event::listen('rainlab.user.activate', function ($user) {
  $group = 
    \Rainlab\User\Models\UserGroup::where('code', 'seller')->first();
     

        $user->groups()->add($group);
    $user->save(); 
}); 
}

elseif  (\Request::is('buyer-registration')){

Event::listen('rainlab.user.activate', function ($user) {
  $group = 
    \Rainlab\User\Models\UserGroup::where('code', 'buyer')->first();
     

        $user->groups()->add($group);
    $user->save(); 
}); 
}

}
```

**Comment by umahatokula** [comment]
> Thanks guys for putting me in the right direction. Finally got it to work using this just in case somebody else needs it.
> 
> ```
>   public function boot()
> { 
> 
> if (\Request::is('seller-registration')){
> 
> Event::listen('rainlab.user.activate', function ($user) {
>   $group = 
>     \Rainlab\User\Models\UserGroup::where('code', 'seller')->first();
>      
> 
>         $user->groups()->add($group);
>     $user->save(); 
> }); 
> }
> 
> elseif  (\Request::is('buyer-registration')){
> 
> Event::listen('rainlab.user.activate', function ($user) {
>   $group = 
>     \Rainlab\User\Models\UserGroup::where('code', 'buyer')->first();
>      
> 
>         $user->groups()->add($group);
>     $user->save(); 
> }); 
> }
> 
> }
> ```

How do you connect the plugin where this code is placed to the user activation link?

**Comment by Corymillz** [comment]
> > Thanks guys for putting me in the right direction. Finally got it to work using this just in case somebody else needs it.
> > ```
> >   public function boot()
> > { 
> > 
> > if (\Request::is('seller-registration')){
> > 
> > Event::listen('rainlab.user.activate', function ($user) {
> >   $group = 
> >     \Rainlab\User\Models\UserGroup::where('code', 'seller')->first();
> >      
> > 
> >         $user->groups()->add($group);
> >     $user->save(); 
> > }); 
> > }
> > 
> > elseif  (\Request::is('buyer-registration')){
> > 
> > Event::listen('rainlab.user.activate', function ($user) {
> >   $group = 
> >     \Rainlab\User\Models\UserGroup::where('code', 'buyer')->first();
> >      
> > 
> >         $user->groups()->add($group);
> >     $user->save(); 
> > }); 
> > }
> > 
> > }
> > ```
> 
> How do you connect the plugin where this code is placed to the user activation link?

If i understand what you mean, you don't need to connect anything. The code only fires after a user is activated.

**Comment by kamskanagi** [comment]
hello guys,
i have a similar issue like this. Hope someone can help.

i want to automatically save the user details to another model if they exist a certain usergroup.

**Comment by bennothommo** [comment]
@kamskanagi All the assistance you will need is contained in the comments above. You may also want to look at the [core model events](https://octobercms.com/docs/api/model/beforesave).

**Comment by kamskanagi** [comment]
@bennothommo Thanks for the reply.
i try the comments above but it dont work.

This is what i want to do.

public function boot()
   {
      1. First, check if the user is in marketing groups or guest group
      2. if the user is in marketing groups
          then save the user->id, user->name and user->email into another Model 
           (Acme\Shopping\Models\Sale) . 
  
    }

**Comment by bennothommo** [comment]
@kamskanagi When do you want the transfer of data to the other model to occur though? When the user is created or updated? When the user is registered?

**Comment by kamskanagi** [comment]
@bennothommo When the user is already registered and added to marketing group then i want the transfer of data to occur.

**Comment by bennothommo** [comment]
@kamskanagi You can extend the `RainLab\Users\Models\User` model and listen to the `model.afterSave` event in order to do the transfer. Check out the docs on how to set this up here: https://octobercms.com/docs/database/model#extending-models

**Comment by kamskanagi** [comment]
@bennothommo Thanks

**Comment by joelfriesecke** [comment]
I understood that i was using "$user->groups()->add($group)" i can add a group to a user,
but how can i remove the groups again automatically if i delete the User?

$user->groups()->remove(); or $user->groups()->delete(); not working
Thanks for helping!

**Comment by bennothommo** [comment]
@joelfriesecke Please post this as another issue, as opposed to hijacking old issues.
